types/host.go

```diff
@@ -0,0 +1,368 @@
+	RPCAccountBalanceCost Currency `json:"rpcAccountBalanceCost"`
+	// FundAccountCost is the cost of funding an ephemeral account on the
+	// host.
+	RPCFundAccountCost Currency `json:"rpcFundAccountCost"`
+	// LatestRevisionCost is the cost of asking the host for the latest
+	// revision of a contract.
+	RPCLatestRevisionCost Currency `json:"rpcLatestRevisionCost"`
+	// RPCRenewContractCost is the cost of renewing a contract with the host.
+	RPCRenewContractCost Currency `json:"rpcRenewContractCost"`
+	// RPCHostSettingsCost is the cost of fetching a new price table from the
+	// host.
+	RPCHostSettingsCost Currency `json:"rpcHostSettingsCost"`
+
+	// ProgInitBaseCost is the cost in Hastings that is incurred when an MDM
+	// program starts to run. This doesn't include the memory used by the
+	// program data. The total cost to initialize a program is calculated as
+	// InitCost = InitBaseCost + MemoryTimeCost * Time
+	ProgInitBaseCost Currency `json:"progInitBaseCost"`
+	// ProgMemoryTimeCost is the cost in Hastings per byte per time that is
+	// incurred by the memory consumption of the program.
+	ProgMemoryTimeCost Currency `json:"progMemorytimecost"`
+	// ProgReadCost is the cost in Hastings per byte of data read from disk
+	// during program executions.
+	ProgReadCost Currency `json:"progReadCost"`
+	// ProgWriteCost is the cost in Hastings per byte of data written to disk
+	// during program execution.
+	ProgWriteCost Currency `json:"progWriteCost"`
+
+	// InstrDropSectorsBaseCost is the base cost in Hastings to execute a drop
+	// sectors instruction.
+	InstrDropSectorsBaseCost Currency `json:"instrDropSectorsBaseCost"`
+	// InstrDropSectorsUnitCost is the additional cost in Hastings per sector of
+	// a drop sectors instruction.
+	InstrDropSectorsUnitCost Currency `json:"instrDropSectorsUnitCost"`
+	// InstrHasSectorBaseCost is the base amount in Hastings to execute a has
+	// sector instruction.
+	InstrHasSectorBaseCost Currency `json:"instrHasSectorBaseCost"`
+	// InstrReadBaseCost is the base cost in Hastings to execute a read
+	// instruction.
+	InstrReadBaseCost Currency `json:"instrReadBaseCost"`
+	// InstrSwapSectorBaseCost is the base cost of swapping two sectors in a
+	// contract.
+	InstrSwapSectorBaseCost Currency `json:"instrSwapSectorCost"`
+	// InstrRevisionBaseCost is the base cost in Hastings to execute a contract
+	// revision instruction.
+	InstrRevisionBaseCost Currency `json:"instrRevisionBaseCost"`
+	// InstrWriteBaseCost is the base cost in Hastings to execute a write
+	// instruction.
+	InstrWriteBaseCost Currency `json:"instrWritebasecost"`
+	// InstrReadRegistryBaseCost is the base cost in Hastings to execute a read
+	// registry instruction.
+	InstrReadRegistryBaseCost Currency `json:"instrReadRegistryBaseCost"`
+	// InstrUpdateRegistryBaseCost is the base cost in Hastings to execute an
+	// update registry instruction.
+	InstrUpdateRegistryBaseCost Currency `json:"instrUpdateRegistryBaseCost"`
```

Comment:
Doing `val, ok := settings.Get(rhp.SettingMaxCollateral)` doesn't seem very pleasant, but having a more dynamic settings structure could be nice. We could also consider binary encoding it for `RPCSettingsResponse` instead of JSON.

Reply:
I imagine you'd only need to check for support once, and after that you can access the field directly:
```go
if _, ok := settings.Prices[rpc.InstrUpdateRegistry]; !ok {
    return errors.New("operation not supported")
}
cost := settings.Prices[rpc.InstrUpdateRegistry].Mul64(foo) // etc.
```

The settings were JSON encoded so that they could be extended later without breaking old renters, but we haven't made much use of that capability. A binary-encoded dynamic map might be a good compromise between efficiency and flexibility.